### PR TITLE
gh-125604: Move _Py_AuditHookEntry, etc. Out of pycore_runtime.h

### PR DESCRIPTION
**Pull request by ericsnowcurrently**
This is essentially a cleanup, moving a handful of API declarations to the header files where they fit best, creating new ones when needed.

We do the following:

* add pycore_debugger_utils.h and move `_Py_DebugOffsets`, etc. there
* inline `struct _getargs_runtime_state` and `struct _gilstate_runtime_state` in `_PyRuntimeState`
* move `struct _reftracer_runtime_state` to the existing pycore_object_state.h
* add pycore_audit.h and move to it `_Py_AuditHookEntry `, `_PySys_Audit()`, and `_PySys_ClearAuditHooks`
* add audit.h and cpython/audit.h and move the existing audit-related API there
* move the perfmap/trampoline API from cpython/sysmodule.h to cpython/ceval.h, and remove the now-empty cpython/sysmodule.h

@pablogsal, I'd love to know if you're okay with us moving `_Py_DebugOffsets` like I have.  Would it break anything?  Would the separate header be inconvenient for external tools or will it make no difference?  Also, do you think the name of the new header is okay?

@zooba, are you okay with us moving *all* the audit API into the new *audit.h header files like I have?

* Issue: gh-125604

### CODE REVIEW

**Comment by ericsnowcurrently** [comment]
FYI, I was considering adding in the small `GENERATE_DEBUG_SECTION()` refactor @1st1 does in gh-124640, putting it in the new pycore_debugger_utils.h, but figured we can do that separately (or keep it in gh-124640.

**Comment by pablogsal** [comment]
> @pablogsal, I'd love to know if you're okay with us moving `_Py_DebugOffsets` like I have. Would it break anything? Would the separate header be inconvenient for external tools or will it make no difference? Also, do you think the name of the new header is okay?

I am fine with moving the definition to a separate header, this is also more convenient for tools to copy-paste. :)



> FYI, I was considering adding in the small `GENERATE_DEBUG_SECTION()` refactor @1st1 does in [gh-124640](https://github.com/python/cpython/pull/124640), putting it in the new pycore_debugger_utils.h, but figured we can do that separately (or keep it in [gh-124640](https://github.com/python/cpython/pull/124640).

I was planning to do this independently if https://github.com/python/cpython/pull/124640 (I did that change in the PR 😉 ) takes time to land so I will build on top of this.

**Comment by ericsnowcurrently** [comment]
> this is also more convenient for tools to copy-paste. :)

I had a gut feeling something like that was the case. :)

**Comment by zooba** [comment]
Provided the audit machinery doesn't gain accessible entry points (which basically means they all exist in the one source unit, via headers is okay), I don't mind where it lives.

If the functions or structure layouts are discoverable at runtime, it makes it much easier to create a bypass. We call an internal-only  version of the function so that it's harder to shim it. Just don't make that any more accessible.

**Comment by ericsnowcurrently** [comment]
I'll merge this tomorrow, unless there are any objections.  If this is blocking anyone, feel free to merge it before I get to it.

File: Include/internal/pycore_debugger_utils.h [hunk 1]
```diff
@@ -0,0 +1,269 @@
+#ifndef Py_INTERNAL_DEBUGGER_UTILS_H
```
Comment (vstinner) [inline Include/internal/pycore_debugger_utils.h]: I suggest to rename the file `pycore_debug_offsets.h` instead. "debug utils" is vague.

Comment (ericsnowcurrently) [reply Include/internal/pycore_debugger_utils.h]: I was going to point out that there are other debugger-related things that would go there, but your suggestion looks very correct to me. 😄  Thanks!  I'll change it.

Comment (ericsnowcurrently) [reply Include/internal/pycore_debugger_utils.h]: done

**Review by vstinner — APPROVED** [review]
LGTM

**Comment by Coffigny94** [comment]
Message Encryption by Microsoft Office 365

Jose A Coffigny Hernandez ***@***.***) has sent you an encrypted message.
Read the message


Learn about messages protected by Microsoft Purview Message Encryption.
Microsoft respects your privacy. To learn more, please read our Privacy Statement.
Microsoft Corporation, One Microsoft Way, Redmond, WA 98052

**Comment by Coffigny94** [comment]
please


________________________________________

**Pull request merged**